### PR TITLE
Size Legend Support for DotplotGL

### PR DESCRIPTION
**Pull request by OssamaRafique**
## Introducing Size Legend for DotplotGL

In this update, I've integrated a size legend feature to `DotplotGL` This addition aids users in deciphering varying dot sizes corresponding to data magnitudes.

### How to Implement the Size Legend

To initialize the size legend on your `DotplotGL` visualization:

1. Feed the appropriate data via the `setState` function:

```
plot.setState({
  sizeLegendData: {
    "minSize": 1,
    "maxSize": 10,
    "steps": 5,
  },
  // Additional data
});
```

2. Configure the legend by invoking the `setSizeLegendOptions` method:

```
dotplotInstance.setSizeLegendOptions(legendOptions, legendDomElement);
```

#### Detailed Parameter Insights:

- `legendOptions`: An object encompassing several attributes to fine-tune the size legend's aesthetics and placement:
  - `orientation`: Specifies the legend's orientation. Valid inputs are: 'horizontal', 'horizontal-inverted', 'vertical', and 'vertical-inverted'. Defaults to 'horizontal'.
  - `position`: Controls the legend's corner placement on the canvas. Options include 'top-left', 'top-right', 'bottom-left', and 'bottom-right'. The standard setting is 'top-right'.
  - `circleColor`: Designates the color for the size circles. The conventional color is 'gray'.
  - `fontSize`: Determines the font size for the legend annotations. The typical size is '12px'.
  - `fontColor`: Sets the text color for the legend labels. The default color is 'black'.
  - `svgPadding`: Specifies the padding around the legend. The default value is derived from the `DEFAULT_SIZE_LEGEND_SVG_PADDING` constant.
  - `circleGap`: Controls the space between circles in the legend. The default value is obtained from the `DEFAULT_SIZE_LEGEND_CIRCLE_GAP` constant.
  - `circleTextGap`: Specifies the gap between circles and their associated text in the legend. The default value is sourced from the `DEFAULT_SIZE_LEGEND_CIRCLE_TEXT_GAP` constant.
  
 Example: 
image

### CODE REVIEW

File: src/DotplotGL.js [hunk 1]
```diff
@@ -97,6 +140,356 @@ class DotplotGL extends BaseGL {
...
+    if (isOrientationHorizontal) {
+      this.sizeLegendSvgNode.attr(
+        "width",
+        circleGroupBBox.width + this.sizeLegendOptions.svgPadding * 2
+      );
+      this.sizeLegendSvgNode.attr(
+        "height",
+        circleGroupBBox.height +
+          textGroupBBox.height +
+          +this.sizeLegendOptions.circleTextGap +
```
Comment (zoecastillo) [inline src/DotplotGL.js]: Is the extra + needed here?

**Comment by zoecastillo** [comment]
General question: What is the legend for the size of the dots supposed to tell us? If I understand the code correctly, the legend text is calculated based on the size of the dot? 
What determines the size of the dot and is that data relevant to the use case? If it is, wouldn't we want the legend to show the data not the size?

File: src/DotplotGL.js (hunk 1 above)
Comment (OssamaRafique) [reply src/DotplotGL.js]: Sorry, Must be a typo

**Comment by OssamaRafique** [comment]
> General question: What is the legend for the size of the dots supposed to tell us? If I understand the code correctly, the legend text is calculated based on the size of the dot? What determines the size of the dot and is that data relevant to the use case? If it is, wouldn't we want the legend to show the data not the size?

Yeah, your comment, makes sense. I thought of it like a representation of size value. But to the users size value may not matter, they would want to see some useful info.

cc: @jkanche What are your thoughts?

**Comment by jkanche** [comment]
yes the size legend should show map the size of the dot to the value in the data it corresponds to not the size itself.

**Comment by jkanche** [comment]
is this still in draft? is so, please change this

**Comment by OssamaRafique** [comment]
> is this still in draft? is so, please change this

Yes I am working on this, Just finalizing in few hours.

**Comment by zoecastillo** [comment]
The code looks fine to me, but I am not sure if this fulfills the requirements that @jkanche  has for the dotplot. Let's wait for his review on this one.

**Pull request merged**